File: packages/cli/go/hieudoanm/src/services/openrouter/openrouter.go
package openrouter

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sort"
	"strings"
	"time"
)

const BaseURL = "https://openrouter.ai/api/v1"

type Model struct {
	ID            string  `json:"id"`
	Name          string  `json:"name"`
	Description   string  `json:"description"`
	ContextLength int     `json:"context_length"`
	Pricing       Pricing `json:"pricing"`
}

type Pricing struct {
	Prompt     string `json:"prompt"`
	Completion string `json:"completion"`
	Request    string `json:"request"`
}

type modelsResponse struct {
	Data []Model `json:"data"`
}

func FetchFreeModels() ([]Model, error) {
	resp, err := http.Get(BaseURL + "/models")
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("OpenRouter returned HTTP %d", resp.StatusCode)
	}

	var payload modelsResponse
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		return nil, fmt.Errorf("decode error: %w", err)
	}

	var free []Model
	for _, m := range payload.Data {
		if isFree(m.Pricing) {
			free = append(free, m)
		}
	}

	sort.Slice(free, func(i, j int) bool {
		return free[i].ID < free[j].ID
	})

	return free, nil
}

func isFree(p Pricing) bool {
	return (p.Prompt == "0" || p.Prompt == "") &&
		(p.Completion == "0" || p.Completion == "")
}

func ResolveModel(query string, models []Model) *Model {
	q := strings.ToLower(query)

	for i, m := range models {
		if strings.ToLower(m.ID) == q {
			return &models[i]
		}
	}

	for i, m := range models {
		if strings.ToLower(m.ID) == q+":free" {
			return &models[i]
		}
	}

	var idMatches []Model
	for _, m := range models {
		if strings.Contains(strings.ToLower(m.ID), q) {
			idMatches = append(idMatches, m)
		}
	}
	if len(idMatches) == 1 {
		return &idMatches[0]
	}
	if len(idMatches) > 1 {
		sort.Slice(idMatches, func(i, j int) bool {
			return len(idMatches[i].ID) < len(idMatches[j].ID)
		})
		return &idMatches[0]
	}

	var nameMatches []Model
	for _, m := range models {
		if strings.Contains(strings.ToLower(m.Name), q) {
			nameMatches = append(nameMatches, m)
		}
	}
	if len(nameMatches) > 0 {
		sort.Slice(nameMatches, func(i, j int) bool {
			return len(nameMatches[i].ID) < len(nameMatches[j].ID)
		})
		return &nameMatches[0]
	}

	return nil
}

// ── Probing ───────────────────────────────────────────────────────────────────

type ProbeStatus int

const (
	StatusOK ProbeStatus = iota
	StatusRateLimited
	StatusRestricted
	StatusError
)

type ProbeResult struct {
	Model   Model
	Status  ProbeStatus
	Message string
	Latency int64
}

func ProbeModel(m Model, apiKey string) ProbeResult {
	payload := map[string]any{
		"model": m.ID,
		"messages": []map[string]any{
			{"role": "user", "content": "hi"},
		},
		"max_tokens": 1,
		"provider": map[string]any{
			"allow_fallbacks": false,
			"data_collection": "allow",
		},
	}

	body, _ := json.Marshal(payload)

	req, err := http.NewRequest(http.MethodPost, BaseURL+"/chat/completions", bytes.NewReader(body))
	if err != nil {
		return ProbeResult{Model: m, Status: StatusError, Message: err.Error()}
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("HTTP-Referer", "https://github.com/freerouter/freerouter")
	req.Header.Set("X-Title", "freerouter")

	start := time.Now()
	resp, err := http.DefaultClient.Do(req)
	elapsed := time.Since(start).Milliseconds()
	if err != nil {
		return ProbeResult{Model: m, Status: StatusError, Message: err.Error()}
	}
	defer resp.Body.Close()

	raw, _ := io.ReadAll(resp.Body)

	switch resp.StatusCode {
	case http.StatusOK:
		return ProbeResult{Model: m, Status: StatusOK, Latency: elapsed}
	case http.StatusTooManyRequests:
		return ProbeResult{Model: m, Status: StatusRateLimited, Message: extractMessage(raw)}
	case http.StatusNotFound:
		return ProbeResult{Model: m, Status: StatusRestricted, Message: extractMessage(raw)}
	default:
		msg := extractMessage(raw)
		if msg == "" {
			msg = fmt.Sprintf("HTTP %d", resp.StatusCode)
		}
		return ProbeResult{Model: m, Status: StatusError, Message: msg}
	}
}

func extractMessage(raw []byte) string {
	var envelope struct {
		Error struct {
			Message string `json:"message"`
		} `json:"error"`
	}
	if err := json.Unmarshal(raw, &envelope); err == nil && envelope.Error.Message != "" {
		msg := envelope.Error.Message
		if len(msg) > 80 {
			msg = msg[:79] + "…"
		}
		return msg
	}
	return ""
}

File: packages/cli/hieudoanm/rust/src/services/openrouter.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::time::Instant;

pub const BASE_URL: &str = "https://openrouter.ai/api/v1";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Pricing {
    pub prompt: String,
    pub completion: String,
    pub request: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Model {
    pub id: String,
    pub name: String,
    pub description: String,
    pub context_length: i64,
    pub pricing: Pricing,
}

#[derive(Debug, Deserialize)]
struct ModelsResponse {
    data: Vec<Model>,
}

pub fn fetch_free_models() -> Result<Vec<Model>> {
    let url = format!("{}/models", BASE_URL);
    let resp = reqwest::blocking::get(&url).context("request failed")?;

    if !resp.status().is_success() {
        anyhow::bail!("OpenRouter returned HTTP {}", resp.status());
    }

    let payload: ModelsResponse = resp.json().context("decode error")?;

    let mut free: Vec<Model> = payload.data.into_iter().filter(|m| is_free(&m.pricing)).collect();
    free.sort_by(|a, b| a.id.cmp(&b.id));

    Ok(free)
}

fn is_free(p: &Pricing) -> bool {
    (p.prompt == "0" || p.prompt.is_empty())
        && (p.completion == "0" || p.completion.is_empty())
}

pub fn resolve_model(query: &str, models: &[Model]) -> Option<Model> {
    let q = query.to_lowercase();

    // Exact match
    for m in models {
        if m.id.to_lowercase() == q {
            return Some(m.clone());
        }
    }

    // Exact match with :free suffix
    let with_free = format!("{}:free", q);
    for m in models {
        if m.id.to_lowercase() == with_free {
            return Some(m.clone());
        }
    }

    // Partial match on ID
    let id_matches: Vec<&Model> = models.iter().filter(|m| m.id.to_lowercase().contains(&q)).collect();
    if id_matches.len() == 1 {
        return Some(id_matches[0].clone());
    }
    if id_matches.len() > 1 {
        let mut sorted = id_matches.clone();
        sorted.sort_by(|a, b| {
            let a_free = a.id.ends_with(":free");
            let b_free = b.id.ends_with(":free");
            if a_free != b_free {
                return b_free.cmp(&a_free);
            }
            a.id.len().cmp(&b.id.len())
        });
        return Some(sorted[0].clone());
    }

    // Partial match on Name
    let name_matches: Vec<&Model> = models.iter().filter(|m| m.name.to_lowercase().contains(&q)).collect();
    if !name_matches.is_empty() {
        let mut sorted = name_matches.clone();
        sorted.sort_by(|a, b| a.id.len().cmp(&b.id.len()));
        return Some(sorted[0].clone());
    }

    None
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ProbeStatus {
    Ok,
    RateLimited,
    Restricted,
    Error,
}

#[derive(Debug, Clone)]
pub struct ProbeResult {
    pub model: Model,
    pub status: ProbeStatus,
    pub message: String,
    pub latency_ms: i64,
}

pub fn probe_model(model: &Model, api_key: &str) -> ProbeResult {
    let payload = serde_json::json!({
        "model": model.id,
        "messages": [
            {"role": "user", "content": "hi"}
        ],
        "max_tokens": 1,
        "provider": {
            "allow_fallbacks": false,
            "data_collection": "allow"
        }
    });

    let url = format!("{}/chat/completions", BASE_URL);
    let client = reqwest::blocking::Client::new();

    let start = Instant::now();
    let resp = match client
        .post(&url)
        .json(&payload)
        .header("Authorization", format!("Bearer {}", api_key))
        .header("HTTP-Referer", "https://github.com/freerouter/freerouter")
        .header("X-Title", "freerouter")
        .send()
    {
        Ok(r) => r,
        Err(e) => {
            return ProbeResult {
                model: model.clone(),
                status: ProbeStatus::Error,
                message: e.to_string(),
                latency_ms: 0,
            }
        }
    };
    let elapsed = start.elapsed().as_millis() as i64;
    let status_code = resp.status().as_u16();
    let body = resp.text().unwrap_or_default();

    match status_code {
        200 => ProbeResult {
            model: model.clone(),
            status: ProbeStatus::Ok,
            message: String::new(),
            latency_ms: elapsed,
        },
        429 => ProbeResult {
            model: model.clone(),
            status: ProbeStatus::RateLimited,
            message: extract_error_message(&body),
            latency_ms: elapsed,
        },
        404 => ProbeResult {
            model: model.clone(),
            status: ProbeStatus::Restricted,
            message: extract_error_message(&body),
            latency_ms: elapsed,
        },
        code => {
            let msg = extract_error_message(&body);
            ProbeResult {
                model: model.clone(),
                status: ProbeStatus::Error,
                message: if msg.is_empty() { format!("HTTP {}", code) } else { msg },
                latency_ms: elapsed,
            }
        }
    }
}

fn extract_error_message(body: &str) -> String {
    if let Ok(envelope) = serde_json::from_str::<serde_json::Value>(body) {
        if let Some(msg) = envelope.get("error").and_then(|e| e.get("message")).and_then(|m| m.as_str()) {
            let mut s = msg.to_string();
            if s.len() > 80 {
                s.truncate(77);
                s.push_str("...");
            }
            return s;
        }
    }
    String::new()
}
